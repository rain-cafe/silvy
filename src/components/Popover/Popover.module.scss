.popover {
    --popover-background-color: var(--silvy-header-color);
    --popover-item-color: var(--silvy-primary-color);

    display: inline-flex;
    position: relative;   
    white-space: nowrap;

    .container {
        position: absolute;
        overflow: hidden;
        display: flex;
        flex-direction: column;
        min-width: 150px;
        background: var(--popover-background-color);
        border-radius: 5px;
        border: 2px solid var(--popover-item-color);
        box-shadow: rgba(0, 0, 0, 0.16) 0px 3px 6px, rgba(0, 0, 0, 0.23) 0px 3px 6px;

        top: calc(100% - 5px);
        left: 50%;
        transform: translateX(-50%);
        opacity: 0;
        transition: 0.15s ease-in-out;
        transition-property: top, opacity;
    }

    &:not(.opened) .container {
        pointer-events: none;
    }

    &.opened .container {
        top: 100%;
        opacity: 1;
    }
} 