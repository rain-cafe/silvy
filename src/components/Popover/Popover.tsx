import React from 'react';
import { Alignment, Actions } from '../../constants/components';
import { classnames } from '../../utils/classnames';
import { useOpened } from '../../utils/element';
import styles from './Popover.module.scss';

export interface PopoverProps {
    children: React.ReactNode;
    toggle: React.ReactNode;
    opened?: boolean;
    alignment?: Alignment;
    action?: Actions;
}

export function Popover({
    children,
    toggle,
    opened: externallyOpened = false,
    alignment = Alignment.CENTER,
    action = Actions.HOVER
}: PopoverProps) {
    const [opened, eventProps] = useOpened(externallyOpened, action, true);

    return (
        <div
            className={classnames(
                styles.popover,
                opened && styles.opened,
                styles[`al-${alignment}`]
            )}
            data-testid='popover'
            {...eventProps}
        >
            <div className={styles.toggle}>
                {toggle}
            </div>
            <div className={styles.container}>
                {children}
            </div>
        </div>
    );
}