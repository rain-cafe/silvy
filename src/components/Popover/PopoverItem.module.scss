.popoverItem {
    --popover-item-color: var(--silvy-primary-color);

    display: flex;
    gap: 5px;
    align-items: center;
    justify-content: space-between;
    padding: 10px;
    font-size: inherit;
    border: transparent;
    background-color: transparent;

    color: var(--popover-item-color);

    transition: 0.15s ease-in-out;
    transition-property: background-color, color;

    &:hover {
        color: var(--popover-background-color);
        background-color: var(--popover-item-color);
        cursor: pointer;
    }
} 