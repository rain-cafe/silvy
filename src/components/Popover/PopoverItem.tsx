import styles from './PopoverItem.module.scss';

export interface PopoverItemProps {
    children: React.ReactNode;
}

export function PopoverItem({
    children
}: PopoverItemProps) {
    return (
        <div className={styles.popoverItem}>
            {children}
        </div>
    );
}