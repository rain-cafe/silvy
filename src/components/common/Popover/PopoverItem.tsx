import { PolymorphicPropsWithoutRef } from 'react-polymorphic-types';
import styles from './PopoverItem.module.scss';

const DEFAULT_ELEMENT = 'button';

export interface PopoverItemOwnProps {
    children: React.ReactNode;
}

export type PopoverItemProps<
    T extends React.ElementType = typeof DEFAULT_ELEMENT
> = PolymorphicPropsWithoutRef<PopoverItemOwnProps, T>;

export function PopoverItem<E extends React.ElementType = typeof DEFAULT_ELEMENT>({
    as,
    children,
    ...extraProps
}: PopoverItemProps<E>) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const Element: React.ElementType = as || DEFAULT_ELEMENT;

    return (
        <Element className={styles.popoverItem} {...extraProps}>
            {children}
        </Element>
    );
}