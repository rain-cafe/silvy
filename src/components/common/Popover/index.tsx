export * from './PopoverItem';
export * from './Popover';