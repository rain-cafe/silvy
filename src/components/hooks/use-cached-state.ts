import { DependencyList, Dispatch, SetStateAction, useEffect, useState } from 'react';

export function useCachedState<T>(supplier: () => T, dependencies: DependencyList): [T, Dispatch<SetStateAction<T>>] {
    const [value, setValue] = useState<T>(supplier());

    useEffect(() => setValue(supplier()), dependencies);

    return [value, setValue];
}

export function useReadOnlyCachedState<T>(supplier: () => T, dependencies: DependencyList): T {
    const [value] = useCachedState(supplier, dependencies);

    return value;
}