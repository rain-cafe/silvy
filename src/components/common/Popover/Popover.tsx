import React from 'react';
import { useRef } from 'react';
import { Alignment, Actions } from '../../../constants/components';
import classnames from 'classnames';
import styles from './Popover.module.scss';
import { useOpened } from '../../hooks/use-opened';

export interface PopoverProps {
    children: React.ReactNode;
    toggle: React.ReactNode;
    opened?: boolean;
    alignment?: Alignment;
    action?: Actions;
}

export function Popover({
    children,
    toggle,
    opened: externallyOpened = false,
    alignment = Alignment.CENTER,
    action = Actions.HOVER
}: PopoverProps) {
    const ref = useRef<HTMLDivElement>();
    const [opened, eventProps] = useOpened({
        opened: externallyOpened,
        action,
        autoClose: true,
        ref
    });

    return (
        <div
            className={classnames(
                styles.popover,
                opened && styles.opened,
                styles[`al-${alignment}`]
            )}
            data-testid='popover'
            ref={ref}
            {...eventProps}
        >
            <div className={styles.toggle}>
                {toggle}
            </div>
            <div className={styles.container}>
                {children}
            </div>
        </div>
    );
}