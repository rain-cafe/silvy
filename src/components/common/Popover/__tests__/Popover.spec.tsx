import { createRenderComponent } from '@rain/test-utils/render';
import { screen } from '@testing-library/react';
import { useOpened } from '../../../hooks/use-opened';
import { Popover, PopoverProps } from '../Popover';
import styles from '../Popover.module.scss';

jest.mock('../../../hooks/use-opened');

const mockedUseOpened = jest.mocked(useOpened);

describe('Component(Popover)', () => {
    const renderPopover = createRenderComponent<PopoverProps>(Popover, {
        children: (
            <>
                <div>Item A</div>
                <div>Item B</div>
            </>
        ),
        toggle: (
            <button>Test</button>
        )
    });

    beforeEach(() => {
        mockedUseOpened.mockReturnValue([false, {}]);
    })

    it('should display the Popover', () => {
        const { baseElement } = renderPopover();

        expect(baseElement).toMatchSnapshot();
    });

    it('should be closed by default', async () => {
        renderPopover({
            href: 'https://silvy.rains.cafe'
        });

        const element = await screen.findByTestId('popover');
        expect(Array.from(element.classList)).not.toContain(styles.opened);
    });

    it('should support being opened', async () => {
        mockedUseOpened.mockReturnValue([true, {}]);

        renderPopover({
            href: 'https://silvy.rains.cafe'
        });

        const element = await screen.findByTestId('popover');
        expect(Array.from(element.classList)).toContain(styles.opened);
    });
});
