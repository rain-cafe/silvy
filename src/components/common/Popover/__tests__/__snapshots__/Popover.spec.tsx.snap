// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(Popover) should display the Popover 1`] = `
<body>
  <div>
    <div
      class="popover al-center"
      data-testid="popover"
    >
      <div
        class="toggle"
      >
        <button>
          Test
        </button>
      </div>
      <div
        class="container"
      >
        <div>
          Item A
        </div>
        <div>
          Item B
        </div>
      </div>
    </div>
  </div>
</body>
`;
